src/core/main.go

```diff
@@ -86,17 +88,27 @@ func updateInitPassword(ctx context.Context, userID int, password string) error
 	return nil
 }
 
-func gracefulShutdown(closing, done chan struct{}) {
+func gracefulShutdown(closing, done chan struct{}, shutdowns ...func()) {
 	signals := make(chan os.Signal, 1)
 	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT)
 	log.Infof("capture system signal %s, to close \"closing\" channel", <-signals)
 	close(closing)
-	select {
-	case <-done:
+	shutdownChan := make(chan struct{}, 1)
+	go func() {
+		for _, s := range shutdowns {
+			s()
```

Comment:
do we need to log the shutdown error?

Reply:
I think error should be handled by shutdown function, the golang HTTP server library has a similar design, its registered shutdown functions do not need to return the error for it.